### PR TITLE
proposal: pipeline

### PR DESCRIPTION
**Pull request by Xieql**
**What type of PR is this?**

/kind design

/kind documentation

**What this PR does / why we need it**:

part of #435

### CODE REVIEW

**Review by hzxuzhonghu — COMMENTED** [review]
I will take a look later

File: docs/proposals/pipeline/pipeline.md [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+
+    Display the logs of the most recent PipelineExecution triggered by an event for a specified Pipeline.
+
+##### 2. PipelineExecution Commands
+
+- **`pipeline-execution list [--pipeline=<name>]`**
+  
+    List all PipelineExecution instances. The list includes each execution's event information and timestamp. 
+    
+    If the `--pipeline` parameter is provided, only list PipelineExecutions belonging to a specific Pipeline.
```
Comment (LiZhenCheng9527) [inline docs/proposals/pipeline/pipeline.md:333]:
```suggestion
    If the `--pipeline` parameter is provided, only PipelineExecutions belonging to a specific Pipeline are listed.
```

File: docs/proposals/pipeline/pipeline.md [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+
+	// SharedWorkspace is the name of the PVC. If not specified, a PVC with the Pipeline's name as prefix will be created by default.
+	// if not set, Kurator will create a pvc named Pipeline.name using default config
+	// +optional
+	SharedWorkspace *string `json:"sharedWorkspace,omitempty"`
+
+	// Results are the list of results written out by the pipeline task's containers
+	// +optional
+	// +listType=atomic
+	Results []tektonv1.PipelineRunResult `json:"results,omitempty"`
```
Comment (hzxuzhonghu) [inline docs/proposals/pipeline/pipeline.md:181]: Suggest we donot its struct directly, we should reduce the dependency on projects. Currently we have a great difficulty to bump dependency

File: docs/proposals/pipeline/pipeline.md [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+
+	Spec   PipelineSpec   `json:"spec"`
+	Status PipelineStatus `json:"status,omitempty"`
+}
+
+// PipelineSpec defines the desired state of a Pipeline.
+type PipelineSpec struct {
+	// Description allows an administrator to provide a description of the pipeline.
+	// +optional
+	Description *string `json:"description,omitempty"`
```
Comment (hzxuzhonghu) [inline docs/proposals/pipeline/pipeline.md:168]: Why need a pointer to string

File: docs/proposals/pipeline/pipeline.md [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+	// Tasks is a list of tasks in the pipeline, containing detailed information about each task.
```
Comment (hzxuzhonghu) [inline docs/proposals/pipeline/pipeline.md:170]: ordered list of tasks, and  add that they will run in order

File: docs/proposals/pipeline/pipeline.md [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+	Tasks []PipelineTask `json:"tasks"`
+
+	// SharedWorkspace is the name of the PVC. If not specified, a PVC with the Pipeline's name as prefix will be created by default.
+	// if not set, Kurator will create a pvc named Pipeline.name using default config
+	// +optional
+	SharedWorkspace *string `json:"sharedWorkspace,omitempty"`
+
+	// Results are the list of results written out by the pipeline task's containers
```
Comment (hzxuzhonghu) [inline docs/proposals/pipeline/pipeline.md:178]:
```suggestion
	// Results are the list of results written out by the pipeline's task
```

File: docs/proposals/pipeline/pipeline.md [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+	// +listType=atomic
+	Env []corev1.EnvVar `json:"env,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,7,rep,name=env"`
+	// ResourceRequirements required by this Step.
+	// Cannot be updated.
+	// More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+	// +optional
+	ResourceRequirements corev1.ResourceRequirements `json:"computeResources,omitempty" protobuf:"bytes,8,opt,name=computeResources"`
+	// Script is the contents of an executable file to execute.
+	//
+	// If Script is not empty, the Step cannot have a Command and the Args will be passed to the Script.
```
Comment (hzxuzhonghu) [inline docs/proposals/pipeline/pipeline.md:256]: `the Step` means the task

File: docs/proposals/pipeline/pipeline.md (hunk 5 above)
Comment (hzxuzhonghu) [reply docs/proposals/pipeline/pipeline.md:178]: The comments is vague, better provide a demo showing how result can be propagated

File: docs/proposals/pipeline/pipeline.md [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+- How will it be tested in isolation vs with other components?
+
+No need to outline all test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+-->
+
+End-to-End Tests: Comprehensive E2E tests should be conducted to ensure the backup, restore, and migration processes work seamlessly across different clusters.
+
+Integration Tests: Integration tests should be designed to ensure Kurator's integration with Velero functions as expected.
```
Comment (LiZhenCheng9527) [inline docs/proposals/pipeline/pipeline.md:359]: Why Velero?

File: docs/proposals/pipeline/pipeline.md [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+
+Unit Tests: Unit tests should cover the core functionalities and edge cases.
+
+Isolation Testing: The backup, restore, and migration functionalities should be tested in isolation and in conjunction with other components to ensure compatibility and performance.
```
Comment (LiZhenCheng9527) [inline docs/proposals/pipeline/pipeline.md:363]: Same problem as above.

File: docs/proposals/pipeline/pipeline.md [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+type PipelineTask struct {
+	// Name is the name of the task.
+	Name string `json:"name"`
+
+	// PredefinedTask is a reference to a predefined task template.
+	// Users should provide a PredefinedTask name from a predefined library.
+	// +optional
+	PredefinedTask *PredefinedTask `json:"predefinedTask,omitempty"`
+
+	// CustomTask enables defining a task directly within the CRD if TaskRef is not used.
```
[the remark this answers is not in the document]
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:193]: IMHO，maybe the origin one is better

File: docs/proposals/pipeline/pipeline.md [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+	// +optional
+	CustomTask *CustomTask `json:"customTask,omitempty"`
+
+	// Retries represents how many times this task should be retried in case of task failure.
+	// +optional
+	Retries *int `json:"retries,omitempty"`
+}
+
+type PredefinedTask struct {
+	// TaskType is used to specify the type of the predefined task.
```
[the remark this answers is not in the document]
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:204]: They are essentially the same.

File: docs/proposals/pipeline/pipeline.md [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+	// +listType=atomic
+	Command []string `json:"command,omitempty" protobuf:"bytes,3,rep,name=command"`
+
+	// Args are the arguments for the entrypoint.
+	// If not provided, the image's CMD is used.
+	// Supports environment variable expansion in the format $(VAR_NAME).
+	// More info: https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/#running-a-command-in-a-shell
+	// +optional
+	// +listType=atomic
+	Args []string `json:"args,omitempty" protobuf:"bytes,4,rep,name=args"`
```
[the remark this answers is not in the document]
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:235]: you mean  `json:"args,omitempty" protobuf:"bytes,4,rep,name=args"`？ actually，they are all the tag to define how the Args field is to be serialized

File: docs/proposals/pipeline/pipeline.md [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+
+##### SLSA Level 2 Key Requirements
+
+1. **Verifiable Signatures**:
+
+    SLSA Level 2 requires verifiable signatures associated with artifacts (like software packages, container images) to ensure their authenticity and integrity.
+
+2. **Complete Build Process Records**:
+
+    Detailed documentation of the entire build process, including inputs and outputs. These records, known as "provenance", provide transparency in artifact creation.
```
[the remark this answers is not in the document]
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:295]: IMHO，they are essentially the same.

File: docs/proposals/pipeline/pipeline.md (hunk 1 above)
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:333]: IMHO，they are essentially the same.

File: docs/proposals/pipeline/pipeline.md (hunk 2 above)
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:181]: I plan to remove the field `Results `, as I've determined it's no longer necessary to expose it to the user.

File: docs/proposals/pipeline/pipeline.md (hunk 3 above)
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:168]: string has a zero value rather than `nil` value 

see 
https://github.com/kubernetes/community/blob/586efe08f05f5bee1ba424ecd53394a16ca54618/contributors/devel/sig-architecture/api-conventions.md?plain=1#L866-L867

File: docs/proposals/pipeline/pipeline.md (hunk 4 above)
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:170]: ok

File: docs/proposals/pipeline/pipeline.md (hunk 5 above)
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:178]: we will remove the `result`, because it is vague for user and not necessary

File: docs/proposals/pipeline/pipeline.md (hunk 7 above)
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:359]: it is a mistake

File: docs/proposals/pipeline/pipeline.md (hunk 6 above)
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:256]: yes, it is `CustomTask`

File: docs/proposals/pipeline/pipeline.md (hunk 3 above)
Comment (hzxuzhonghu) [reply docs/proposals/pipeline/pipeline.md:189]: I dont think we need to distinguish "" and nil. They are same here, as we won't generate a default value for it

Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:189]: ok，

File: docs/proposals/pipeline/pipeline.md [hunk 13]
```diff
@@ -0,0 +1,415 @@
+---
+title: Pipeline in Kurator
+authors:
+- "@Xieql"
+  reviewers:
```
Comment (hzxuzhonghu) [inline docs/proposals/pipeline/pipeline.md:5]: remove indents

File: docs/proposals/pipeline/pipeline.md [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+	// The tasks will be executed in the order they are listed.
+	Tasks []PipelineTask `json:"tasks"`
+
+	// SharedWorkspace is the name of the PVC. If not specified, a PVC with the Pipeline's name as prefix will be created by default.
+	// If not set, Kurator will create a PVC named Pipeline.name using default config
+	// +optional
+	SharedWorkspace *string `json:"sharedWorkspace,omitempty"`
+}
+
+type PipelineTask struct {
```
Comment (hzxuzhonghu) [inline docs/proposals/pipeline/pipeline.md:201]: nit: remove pipeline prefix

File: docs/proposals/pipeline/pipeline.md [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+	// Name is the name of the task.
+	Name string `json:"name"`
+
+	// PredefinedTask is a reference to a predefined task template.
+	// Users should provide a PredefinedTask name from a predefined library.
+	// +optional
+	PredefinedTask *PredefinedTask `json:"predefinedTask,omitempty"`
+
+	// CustomTask enables defining a task directly within the CRD if TaskRef is not used.
+	// This should only be used when TaskRef is not provided.
```
Comment (hzxuzhonghu) [inline docs/proposals/pipeline/pipeline.md:211]: s/TaskRef/PredefinedTask

File: docs/proposals/pipeline/pipeline.md [hunk 16]
```diff
@@ -0,0 +1,415 @@
...
+
+type PipelineStatus struct {
+	// Phase describes the overall state of the Pipeline.
+	// +optional
+	Phase *string `json:"phase,omitempty"`
+
+    // EventListenerServiceName specifies the name of the service created by Kurator for event listeners. 
+    // This name is useful for users when setting up a gateway service and routing to this service.
+    // +optional
+    EventListenerServiceName *string `json:"eventListenerServiceName,omitempty"`
```
Comment (hzxuzhonghu) [inline docs/proposals/pipeline/pipeline.md:289]: indent

File: docs/proposals/pipeline/pipeline.md [hunk 17]
```diff
@@ -0,0 +1,415 @@
...
+type PipelineList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Pipeline `json:"items"`
+}
+```
+
+#### How to use pipeline
+
+![use-pipeline](./image/use-pipeline.svg)
```
[the remark this answers is not in the document]
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:304]: ok

File: docs/proposals/pipeline/pipeline.md (hunk 16 above)
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:289]: I'm not sure what the 'indent' means here

File: docs/proposals/pipeline/pipeline.md [hunk 18]
```diff
@@ -0,0 +1,415 @@
...
+	// Script is the contents of an executable file to execute.
+	// If Script is not empty, the CustomTask cannot have a Command and the Args will be passed to the Script.
+	// +optional
+	Script string `json:"script,omitempty"`
+}
+
+type PipelineStatus struct {
+	// Phase describes the overall state of the Pipeline.
+	// +optional
+	Phase *string `json:"phase,omitempty"`
```
[the remark this answers is not in the document]
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md:284]: yes,we need more details

File: docs/proposals/pipeline/pipeline.md (hunk 16 above)
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md]: fixed

File: docs/proposals/pipeline/pipeline.md [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+
+	// Retries represents how many times this task should be retried in case of task failure.
+	// +optional
+	Retries *int `json:"retries,omitempty"`
+}
+
+type PredefinedTask struct {
+	// TaskType is used to specify the type of the predefined task.
+	// This is a required field and determines which task template will be used.
+	TaskType string `json:"taskType"`
```
[the remark this answers is not in the document]
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md]: fiexd

File: docs/proposals/pipeline/pipeline.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (Xieql) [reply docs/proposals/pipeline/pipeline.md]: more details has been added

**Comment by Xieql** [comment]
@hzxuzhonghu PTAL

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm
